sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c

```diff
@@ -520,77 +520,118 @@ size_t gnrc_sixlowpan_iphc_decode(gnrc_pktsnip_t **dec_hdr, gnrc_pktsnip_t *pkt,
+        nhc_data[nhc_len++] = dst_port - NHC_UDP_4BIT_PORT +
+                              ((src_port - NHC_UDP_4BIT_PORT) << 4);
     }
-    else if ((byteorder_ntohs(*dst_port) & NHC_UDP_8BIT_MASK) == NHC_UDP_8BIT_PORT) {
+    else if ((dst_port & NHC_UDP_8BIT_MASK) == NHC_UDP_8BIT_PORT) {
         DEBUG("6lo iphc nhc: elide dst\n");
-        ipv6_hdr->nh = NHC_UDP_S_INLINE;
-        nhc_len += 2; /* keep src_port */
-        udp_data[nhc_len++] = byteorder_ntohs(*dst_port) - NHC_UDP_8BIT_PORT;
-        udp_data[nhc_len++] = udp_hdr->checksum.u8[0];
-        udp_data[nhc_len++] = udp_hdr->checksum.u8[1];
+        nhc_data[0] |= NHC_UDP_S_INLINE;
+        nhc_data[nhc_len++] = udp_hdr->src_port.u8[0];
+        nhc_data[nhc_len++] = udp_hdr->src_port.u8[1];
+        nhc_data[nhc_len++] = dst_port - NHC_UDP_8BIT_PORT;
     }
-    else if ((byteorder_ntohs(*src_port) & NHC_UDP_8BIT_MASK) == NHC_UDP_8BIT_PORT) {
+    else if ((src_port & NHC_UDP_8BIT_MASK) == NHC_UDP_8BIT_PORT) {
         DEBUG("6lo iphc nhc: elide src\n");
-        ipv6_hdr->nh = NHC_UDP_D_INLINE;
-        udp_data[nhc_len++] = byteorder_ntohs(*src_port) - NHC_UDP_8BIT_PORT;
-        udp_data[nhc_len++] = udp_hdr->dst_port.u8[0];
-        udp_data[nhc_len++] = udp_hdr->dst_port.u8[1];
-        udp_data[nhc_len++] = udp_hdr->checksum.u8[0];
-        udp_data[nhc_len++] = udp_hdr->checksum.u8[1];
+        nhc_data[0] |= NHC_UDP_D_INLINE;
+        nhc_data[nhc_len++] = src_port - NHC_UDP_8BIT_PORT;
+        nhc_data[nhc_len++] = udp_hdr->dst_port.u8[0];
+        nhc_data[nhc_len++] = udp_hdr->dst_port.u8[1];
     }
     else {
         DEBUG("6lo iphc nhc: src and dst inline\n");
-        ipv6_hdr->nh = NHC_UDP_SD_INLINE;
-        nhc_len = sizeof(udp_hdr_t) - 4; /* skip src + dst and elide length */
-        udp_data[nhc_len++] = udp_hdr->checksum.u8[0];
-        udp_data[nhc_len++] = udp_hdr->checksum.u8[1];
+        nhc_data[0] |= NHC_UDP_SD_INLINE;
+        nhc_data[nhc_len++] = udp_hdr->src_port.u8[0];
+        nhc_data[nhc_len++] = udp_hdr->src_port.u8[1];
+        nhc_data[nhc_len++] = udp_hdr->dst_port.u8[0];
+        nhc_data[nhc_len++] = udp_hdr->dst_port.u8[1];
     }
 
-    /* Set UDP header ID (rfc6282#section-5). */
-    ipv6_hdr->nh |= NHC_UDP_ID;
-
-    /* In case payload is in this snip (e.g. a forwarded packet):
-     * move data to right place */
-    size_t diff = sizeof(udp_hdr_t) - nhc_len;
-    for (size_t i = nhc_len; i < (udp->size - diff); i++) {
-      udp_data[i] = udp_data[i + diff];
-    }
-    /* NOTE: gnrc_pktbuf_realloc_data overflow if (udp->size - diff) < 4 */
-    gnrc_pktbuf_realloc_data(udp, (udp->size - diff));
+    /* TODO: Add support for elided checksum. */
+    nhc_data[nhc_len++] = udp_hdr->checksum.u8[0];
+    nhc_data[nhc_len++] = udp_hdr->checksum.u8[1];
 
     return nhc_len;
 }
 #endif
 
+static inline bool _compressible(gnrc_pktsnip_t *hdr)
+{
+    switch (hdr->type) {
+        case GNRC_NETTYPE_UNDEF:    /* when forwarded */
```

Comment:
This feels wrong, in principle. GNRC_NETTYPE_UNDEF could be anything, but here we assume it is an ipv6 header always. If I remember correctly, and looking at the comment, the packet type is not set on forwarded packets.

Reply:
> If I remember correctly, and looking at the comment, the packet type is not set on forwarded packets.

Yes, and why should it? It doesn't make sense to parse a packet to more than the IP header if it is just forwarded to the next hop. I also wouldn't say that this is wrong in the scope of 6Lo alone. While it is not implemented (yet) Generic Header Compression [RFC 7400](https://tools.ietf.org/html/rfc7400) allows for compression of *any* header (i.e. undefined headers as well ;-)).

Reply:
I think I am missing some piece of information, but why would we assume that we can compress anything that is marked as undefined or IPV6, but assume that any other nettypes are incompressible?
If GNRC_NETTYPE_UNDEF is only used for router forwarding, why not rename it to GNRC_NETTYPE_FORWARDED?

Reply:
> I think I am missing some piece of information, but why would we assume that we can compress anything that is marked as undefined or IPV6, but assume that any other nettypes are incompressible.

Because there is currently no definition for it. Sure I could add `GNRC_NETTYPE_TCP` or `GNRC_NETTYPE_ICMPV6` here, but the primary goal here is to keep the `gnrc_pktsnip_t` dispatch as *small* as possible here. Fact is (and that is that is the main if not the only reason why it is included at the moment here), forwarded packets are marked as `GNRC_NETTYPE_UNDEF`, not including them would break 6Lo for multihop scenarios.

> If GNRC_NETTYPE_UNDEF is only used for router forwarding, why not rename it to GNRC_NETTYPE_FORWARDED?

Because that would (a) fuzzy the semantics of `gnrc_nettype_t`, which is there to identify a certain protocol type within a packet (if it is known) and the handlers of that type, (b) adding random nettypes it would bloat `gnrc_netreg` because it uses `GNRC_NETTYPE_NUMOF` to allocate its registry space. Thus, I don't see any benefit to add an extra nettype.

Reply:
Also note that [I only use the first `GNRC_NETTYPE_UNDEF` I find](https://github.com/RIOT-OS/RIOT/pull/8626/files#diff-bb73c1670f25f07e92a84167a3fa0954R623) (which usually is the forwarded payload).